// swrn/dist/lib/app.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _router = require('./router');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var App = function (_React$Component) {
    (0, _inherits3.default)(App, _React$Component);

    function App() {
        (0, _classCallCheck3.default)(this, App);

        var _this = (0, _possibleConstructorReturn3.default)(this, (App.__proto__ || (0, _getPrototypeOf2.default)(App)).call(this));

        _this.state = {
            routerComponents: {}
        };
        return _this;
    }

    (0, _createClass3.default)(App, [{
        key: 'getChildContext',
        value: function getChildContext() {
            var props = (0, _assign2.default)({}, this.props);
            delete props.Main;
            return {
                SWRN_PROPS: (0, _extends3.default)({}, props, {
                    routerComponents: this.state.routerComponents
                })
            };
        }

        //设置popstate事件

    }, {
        key: 'componentDidMount',
        value: function componentDidMount() {
            var _this2 = this;

            var routeLoader = this.props.routeLoader;

            window.addEventListener('popstate', function () {
                var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(e) {
                    var _ResolveRouter, route, Component;

                    return _regenerator2.default.wrap(function _callee$(_context) {
                        while (1) {
                            switch (_context.prev = _context.next) {
                                case 0:
                                    _ResolveRouter = (0, _router.ResolveRouter)(_this2.props.data.routes, location.pathname), route = _ResolveRouter.page;
                                    _context.next = 3;
                                    return routeLoader.loadPage(route);

                                case 3:
                                    Component = _context.sent;

                                    routeLoader.notify({ Component: Component, route: route });

                                case 5:
                                case 'end':
                                    return _context.stop();
                            }
                        }
                    }, _callee, _this2);
                }));

                return function (_x) {
                    return _ref.apply(this, arguments);
                };
            }());
        }
    }, {
        key: 'render',
        value: function render() {
            var Main = this.props.Main;

            return _react2.default.createElement(Main, null);
        }
    }]);
    return App;
}(_react2.default.Component);

exports.default = App;


App.childContextTypes = {
    SWRN_PROPS: _propTypes2.default.object
};
