// swrn/dist/server/build/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _webpackDevMiddleware = require('webpack-dev-middleware');

var _webpackDevMiddleware2 = _interopRequireDefault(_webpackDevMiddleware);

var _router = require('../router');

var _router2 = _interopRequireDefault(_router);

var _webpack = require('./webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _clean = require('./clean');

var _clean2 = _interopRequireDefault(_clean);

var _compiler = require('../compiler');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

global.SWRN_InServer = true;

exports.default = function () {
    var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(dir) {
        var _this = this;

        var conf = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

        var dist, options, routes, _ref2, _ref3, compiler;

        return _regenerator2.default.wrap(function _callee3$(_context3) {
            while (1) {
                switch (_context3.prev = _context3.next) {
                    case 0:
                        dist = '.swrn';
                        options = {
                            dir: dir,
                            dev: false,
                            dist: dist,
                            page: 'pages',
                            assetPrefix: '/swrn'

                            // 获取路由
                        };
                        _context3.next = 4;
                        return new _router2.default(options).routes();

                    case 4:
                        routes = _context3.sent;


                        // webpack编译
                        options.routes = routes;
                        options.dist = dist;
                        options.entry = false;

                        _context3.next = 10;
                        return _promise2.default.all([(0, _webpack2.default)(options), (0, _clean2.default)(options)]);

                    case 10:
                        _ref2 = _context3.sent;
                        _ref3 = (0, _slicedToArray3.default)(_ref2, 1);
                        compiler = _ref3[0];


                        (0, _webpackDevMiddleware2.default)(compiler, {
                            noInfo: true,
                            quiet: true,
                            clientLogLevel: 'warning'
                        });

                        return _context3.abrupt('return', new _promise2.default(function (resolve, reject) {
                            compiler.plugin('done', (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2() {
                                var _ref5, _ref6, compilerBuild;

                                return _regenerator2.default.wrap(function _callee2$(_context2) {
                                    while (1) {
                                        switch (_context2.prev = _context2.next) {
                                            case 0:
                                                _context2.next = 2;
                                                return (0, _compiler.replaceStaticSource)(options);

                                            case 2:
                                                console.log('build done 50%');

                                                _context2.next = 5;
                                                return _promise2.default.all([(0, _webpack.buildWebpack)(options), (0, _clean.cleanBundles)(options)]);

                                            case 5:
                                                _ref5 = _context2.sent;
                                                _ref6 = (0, _slicedToArray3.default)(_ref5, 1);
                                                compilerBuild = _ref6[0];


                                                (0, _webpackDevMiddleware2.default)(compilerBuild, {
                                                    noInfo: true,
                                                    quiet: true,
                                                    clientLogLevel: 'warning'
                                                });

                                                compilerBuild.plugin('done', (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
                                                    return _regenerator2.default.wrap(function _callee$(_context) {
                                                        while (1) {
                                                            switch (_context.prev = _context.next) {
                                                                case 0:

                                                                    options.fileName = 'common.js';
                                                                    _context.next = 3;
                                                                    return (0, _clean.cleanFile)(options);

                                                                case 3:
                                                                    options.fileName = 'app.js';
                                                                    _context.next = 6;
                                                                    return (0, _clean.cleanFile)(options);

                                                                case 6:
                                                                    options.fileName = 'manifest.js';
                                                                    _context.next = 9;
                                                                    return (0, _clean.cleanFile)(options);

                                                                case 9:

                                                                    console.log('build done 100%');

                                                                    resolve();

                                                                case 11:
                                                                case 'end':
                                                                    return _context.stop();
                                                            }
                                                        }
                                                    }, _callee, _this);
                                                })));

                                            case 10:
                                            case 'end':
                                                return _context2.stop();
                                        }
                                    }
                                }, _callee2, _this);
                            })));
                        }));

                    case 15:
                    case 'end':
                        return _context3.stop();
                }
            }
        }, _callee3, this);
    }));

    function build(_x) {
        return _ref.apply(this, arguments);
    }

    return build;
}();
